libsolidity/codegen/YulUtilFunctions.cpp

```diff
@@ -266,6 +267,66 @@ std::string YulUtilFunctions::requireOrAssertFunction(bool _assert, Type const*
 	});
 }
 
+std::string YulUtilFunctions::requireWithError(FunctionCall const& errorConstructorCall)
+{
+	ErrorDefinition const* errorDefinition = dynamic_cast<ErrorDefinition const*>(ASTNode::referencedDeclaration(errorConstructorCall.expression()));
+	solAssert(errorDefinition);
+
+	std::string const errorSignature = errorDefinition->functionType(true)->externalSignature();
+	std::string const signatureHash = formatNumber(util::selectorFromSignatureU256(errorSignature));
+	auto functionName = [&]() -> std::string {
+		std::string name = "require_helper_t_error_" + std::to_string(errorDefinition->id());
+		for (ASTPointer<Expression const> const& argument: errorConstructorCall.sortedArguments())
+		{
+			solAssert(argument->annotation().type);
+			name += ("_" + argument->annotation().type->identifier());
+		}
+		return name;
+	};
+
+	auto errorArgumentVarsAndTypes = [&]() -> std::tuple<std::vector<std::string>, std::vector<Type const*>> {
+		std::vector<ASTPointer<Expression const>> const& errorConstructorArguments = errorConstructorCall.sortedArguments();
+		std::vector<std::string> errorArgumentVars{};
+		std::vector<Type const*> errorArgumentTypes{};
+		for (ASTPointer<Expression const> const& arg: errorConstructorArguments)
+		{
+			solAssert(arg->annotation().type);
+			if (arg->annotation().type->sizeOnStack() > 0)
+				errorArgumentVars += IRVariable(*arg).stackSlots();
+			errorArgumentTypes.push_back(arg->annotation().type);
+		}
+		return std::tie(errorArgumentVars, errorArgumentTypes);
+	};
+
+	return m_functionCollector.createFunction(functionName(), [&]() {
+		std::vector<Type const*> const& parameterTypes = errorDefinition->functionType(true)->parameterTypes();
+		auto [errorArgumentVars, errorArgumentTypes] = errorArgumentVarsAndTypes();
+		std::string const encodeFunc = ABIFunctions(m_evmVersion, m_revertStrings, m_functionCollector)
+			.tupleEncoder(
+				errorArgumentTypes,
+				parameterTypes
+			);
+
+		return Whiskers(R"(
+			function <functionName>(condition <errorArguments>) {
+				if iszero(condition) {
+					let memPtr := <allocateUnbounded>()
+					mstore(memPtr, <errorHash>)
+					let end := <abiEncodeFunc>(add(memPtr, <hashHeaderSize>) <errorArguments>)
+					revert(memPtr, sub(end, memPtr))
```

Comment:
> Would be best to avoid duplicating this logic and instead reuse revertWithError().

I kinda hate that you're probably right here :) For me, it's 'nicer' this way because I can see what it generates in one place without having to switch to another function and then try to piece things together, but I guess that's my inexperience showing.

As for inlining - `requireOrAssertFunction` also wraps the functionality in a function; @ekpyron should this be inlined or should it stay in a function?

Reply:
This was extensively discussed on the chat and calls, so just to summarize how things stand:
- The fact that the new code is wrapped in a function while `revertWithReason()` is not is not an inconsistency after all. The function covers not only the revert but also the condition. We can technically consider it inlined into that condition.
- `revertWithReason()` should still be reused so that there's only one path generating this code. The duplicated piece of Yul is small, but the point is that having it in two places increases the risk of them diverging accidentally. It could happen e.g. if we merged #14186 after this PR - there would be no conflict.
- We decided that @nikola-matic will take over #14186 and I'll review it.

Reply:
`revertWithError()` you mean?

Reply:
Yeah. Copied the wrong one. Fixed.